include/swift/SIL/Projection.h

```diff
@@ -411,7 +411,8 @@ class ProjectionPath {
   }
 
   ProjectionPath &operator=(ProjectionPath &&O) {
-    *this = std::move(O);
+    if (this != &O)
+      *this = std::move(O);
     O.Path.clear();
```

Comment:
> On Dec 16, 2015, at 1:20 PM, Joe Groff <notifications@github.com> wrote:
> In include/swift/SIL/Projection.h <https://github.com/apple/swift/pull/605#discussion_r47832680>:
> 
> > @@ -411,7 +411,8 @@ class ProjectionPath {
> >    }
> >  
> >    ProjectionPath &operator=(ProjectionPath &&O) {
> > -    *this = std::move(O);
> > +    if (this != &O)
> > +      *this = std::move(O);
> >      O.Path.clear();
> You shouldn't clear the value if this == &O either.
> 
> 

Clang is probably complaining about this because it’s an infinite recursion, not because it’s problematic when source == target.  Self-move-assignment is not a meaningful operation, and most of our move assignment operators don’t deal with it well.

John.